src/IO/S3/deleteFileFromS3.cpp

```diff
@@ -0,0 +1,235 @@
+#include <IO/S3/deleteFileFromS3.h>
+
+#if USE_AWS_S3
+
+#include <Common/logger_useful.h>
+#include <IO/S3/Client.h>
+#include <IO/S3/Requests.h>
+#include <IO/S3/BlobStorageLogWriter.h>
+#include <IO/S3/S3Capabilities.h>
+#include <IO/S3/getObjectInfo.h>
+
+
+namespace ProfileEvents
+{
+    extern const Event S3DeleteObjects;
+}
+
+
+namespace DB
+{
+
+void deleteFileFromS3(
+    const std::shared_ptr<const S3::Client> & s3_client,
+    const String & bucket,
+    const String & key,
+    bool if_exists,
+    BlobStorageLogWriterPtr blob_storage_log,
+    const String & local_path_for_blob_storage_log,
+    size_t file_size_for_blob_storage_log,
+    std::optional<ProfileEvents::Event> profile_event)
+{
+    ProfileEvents::increment(ProfileEvents::S3DeleteObjects);
+    if (profile_event && *profile_event != ProfileEvents::S3DeleteObjects)
+        ProfileEvents::increment(*profile_event);
+
+    auto log = getLogger("deleteFileFromS3");
+
+    S3::DeleteObjectRequest request;
+    request.SetBucket(bucket);
+    request.SetKey(key);
+    auto outcome = s3_client->DeleteObject(request);
+
+    if (blob_storage_log)
+    {
+        blob_storage_log->addEvent(BlobStorageLogElement::EventType::Delete,
+                                   bucket, key,
+                                   local_path_for_blob_storage_log, file_size_for_blob_storage_log,
+                                   outcome.IsSuccess() ? nullptr : &outcome.GetError());
+    }
+    else
+    {
+        LOG_TRACE(log, "No blob storage log, not writing blob {}", key);
+    }
+
+    if (!outcome.IsSuccess() && (!if_exists || !S3::isNotFoundError(outcome.GetError().GetErrorType())))
+    {
+        /// In this case even if absence of key may be ok for us, the log will be polluted with error messages from aws sdk.
+        /// Looks like there is no way to suppress them.
+        const auto & err = outcome.GetError();
+        throw S3Exception(err.GetErrorType(), "{} (Code: {}) while removing object with path {} from S3",
+                          err.GetMessage(), static_cast<size_t>(err.GetErrorType()), key);
+    }
+
+    LOG_DEBUG(log, "Object with path {} was removed from S3", key);
+}
+
+
+void deleteFilesFromS3(
+    const std::shared_ptr<const S3::Client> & s3_client,
+    const String & bucket,
+    const Strings & keys,
+    bool if_exists,
+    std::optional<bool> is_batch_delete_supported,
+    std::function<void(bool)> set_is_batch_delete_supported,
+    size_t batch_size,
+    BlobStorageLogWriterPtr blob_storage_log,
+    const Strings & local_paths_for_blob_storage_log,
+    const std::vector<size_t> & file_sizes_for_blob_storage_log,
+    std::optional<ProfileEvents::Event> profile_event)
+{
+    if (keys.empty())
+       return; /// Nothing to delete.
```

Comment:
In my opinion, you do not need all this logic. Just add try_batch_requests = false if keys.size() <= 1